echopop/computation/biology.py

```diff
@@ -104,160 +128,751 @@ def sum_strata_weight( haul_data: pd.DataFrame ,
+    )
+
+    # ---- Proportions of each sex-length bin pair relative to `stratum_number_all`
+    proportions_unaged_length[ 'proportion_number_all' ] = (
+        proportions_unaged_length.number_all / proportions_unaged_length.stratum_number_all
+    ).fillna( 0 )
+
+    ### Filter out unnecessary columns and return output
+    return proportions_unaged_length.filter( regex = '^(?!number_|stratum_number_).*' )
+
+def compute_unaged_weight_proportions( proportions_unaged_length: pd.DataFrame ,
+                                       length_weight_df: pd.DataFrame ):
+    """
+   Compute the unaged weight proportions across all fish and specific sexes for each
+   length bin  
+
+    Parameters
+    ----------
+    proportions_unaged_length: pd.DataFrame
+        Dataframe containing the total number proportions distributed across
+        each length bin
+    length_weight_df: pd.DataFrame
+        Dataframe containing the modeled weights for each sex and length bin
+        fit from the length-weight regression (fit for all animals)
+    """   
+
+    ### Extract length-weight regression results calculated for all animals
+    length_weight_all = length_weight_df[ length_weight_df.sex == 'all' ][ [ 'length_bin' , 'weight_modeled' ] ]
+
+    ### Calculate the weight proportion of each length bin that is 'weighted' by the number proportions
+    # ---- Merge `proportions_unaged_length_sex` and `length_weight_all`
+    proportions_unaged_weight_length = pd.merge( proportions_unaged_length ,
+                                                 length_weight_all ,
+                                                 on = [ 'length_bin' ] ,
+                                                 how = 'left' )
     
-    ### Calculate interpolated weights based on length bins for each sex per haul 
-    length_weight_fit = (
-        length_weight_df
-        .copy( )
-        .assign( length_bin_value  = lambda x: x[ 'length_bin' ].apply( lambda y: y.mid ) )
+    # ---- Calculate the weight proportion (`w_ln_all_array` in the original Matlab code)
+    proportions_unaged_weight_length[ 'weight_proportion_all' ] = (
+        proportions_unaged_weight_length.weight_modeled * proportions_unaged_weight_length.proportion_number_all
+    )
+
+    ### Sum weight proportions for each strata to get the weight per length bin
+    # ---- Calculate the summed weights per stratum (`w_ln_array_sum` in the original Matlab code)
+    proportions_unaged_weight_length[ 'weight_stratum_proportion' ] = (
+       proportions_unaged_weight_length.groupby( [ 'stratum_num' ] )[ 'weight_proportion_all' ].transform( sum )
     )
+    
+    # ---- Calculate the weight per length distribution bin (`w_ln_all_N` in the original Matlab code)
+    proportions_unaged_weight_length[ 'proportion_weight_length' ] = (
+        proportions_unaged_weight_length.weight_proportion_all / proportions_unaged_weight_length.weight_stratum_proportion
+    )
```

Comment:
Some suggestions to avoid confusion of the quantities:
```suggestion
    # ---- Calculate the estimated weight for each length bin within each stratum (`w_ln_all_array` in the original Matlab code)
    proportions_unaged_weight_length[ 'weight_proportion_all' ] = (
        proportions_unaged_weight_length.weight_modeled * proportions_unaged_weight_length.proportion_number_all
    )

    ### Calculate weight proportions for each length bin for all strata
    # ---- Calculate the summed estimated weights across all length bins within each stratum (`w_ln_array_sum` in the original Matlab code)
    proportions_unaged_weight_length[ 'weight_stratum_proportion' ] = (
       proportions_unaged_weight_length.groupby( [ 'stratum_num' ] )[ 'weight_proportion_all' ].transform( sum )
    )
    
    # ---- Calculate the weight proportion of each length bin within each stratum (`w_ln_all_N` in the original Matlab code)
    proportions_unaged_weight_length[ 'proportion_weight_length' ] = (
        proportions_unaged_weight_length.weight_proportion_all / proportions_unaged_weight_length.weight_stratum_proportion
    )
```